### PR TITLE
Fix which behavior when passed an empty string

### PR DESCRIPTION
**Pull request by tgflynn**
Currently when passed an argument that is the empty string Sys.which() returns the path of a directory.

For example:

```
julia> Sys.which("")
"/usr/local/sbin"
```
This is not correct since the empty string is not a valid path component on any OS of which I am aware.  

This PR fixes this issue so that Sys.which("") returns nothing in accord with its documentation.

### CODE REVIEW

File: base/sysinfo.jl [hunk 1]
```diff
@@ -459,6 +459,9 @@ for executable permissions only (with `.exe` and `.com` extensions added on
 Windows platforms); no searching of `PATH` is performed.
 """
 function which(program_name::String)
+    if isempty(program_name)
```
Comment (staticfloat) [inline base/sysinfo.jl]: Let's nip things like `" "` in the bud as well; do `isempty(strip(program_name))`

File: test/sysinfo.jl [hunk 2]
```diff
@@ -6,3 +6,6 @@
 sprint(Base.Sys.cpu_summary)
 @test Base.Sys.uptime() > 0
 Base.Sys.loadavg()
+
+# Check that which behaves correctly when passed an empty string
+@test isnothing(Base.Sys.which(""))
```
Comment (staticfloat) [inline test/sysinfo.jl]: Let's test `isnothing(Base.Sys.which(" "))` as well

**Review by staticfloat — CHANGES_REQUESTED** [review]

File: base/sysinfo.jl (hunk 1 above)
Comment (tgflynn) [reply base/sysinfo.jl]: So it turns out that this already works with the current code (at least on Linux).  

The reason is that ```isexecutable("/usr/bin/ ") == false``` while ```isexecutable("/usr/bin/") == true```.

It makes sense to add the test though.

Comment (staticfloat) [reply base/sysinfo.jl]: In that case, I think it makes more sense to change `isexecutable()` to `isfile() && isexecutable()`.  What do you think?

Comment (tgflynn) [reply base/sysinfo.jl]: Yes, I agree that makes sense.

I think we might as well keep the early exit for an empty string though.  There's no point in searching a bunch of directories for something they can't possibly contain.

Comment (vtjnash) [reply base/sysinfo.jl]: While "" isn't a valid file name, " " is perfectly legal (if a bit insane) as an executable name.

Comment (tgflynn) [reply base/sysinfo.jl]: Yep, though I think it's OK to include a test for that because the odds of there being such a file on a test system is very low.

There are currently no other tests for Sys.which and in general it's a bit complicated to write them because it would require controlling environment variables as well as the filesystem.  However testing that Sys.which(" ") == nothing is easy and almost certain to be right.

Comment (staticfloat) [reply base/sysinfo.jl]: FWIW, it's not so hard to make bulletproof tests for this:

```
mktempdir() do temp_dir
	my_bin = joinpath(temp_dir, "my_bin")
	touch(my_bin)
	chmod(my_bin, 0o755)
	withenv("PATH" => temp_dir) do
		@test abspath(Sys.which("my_bin")) == abspath(my_bin)
		@test Sys.which(" ") == nothing
		@test Sys.which("") == nothing
	end
end
```

Something like that should work, modulo `chmod()` issues on Windows.  (E.g. don't try to do test creating non-executable files on windows.... everything is executable on windows from Julia's perspective ;) )

Comment (vtjnash) [reply base/sysinfo.jl]: We already have tests for that (in test/spawn.jl), we just need tests for the new code (empty string case).

Comment (staticfloat) [reply base/sysinfo.jl]: Great.  Good to merge from my perspective then.

**Review by staticfloat — APPROVED** [review]

File: base/sysinfo.jl (hunk 1 above)
Comment (tgflynn) [reply base/sysinfo.jl]: @vtjnash 

I'm not clear on why the tests are there.  I thought the convention was that functions in base/foo.jl were tested in test/foo.jl.  

But since that's where the other which tests are I should probably move the new ones into test/spawn.jl which will also allow using the existing testing framework to ensure that the blank filename test is correct.

**Comment by tgflynn** [comment]
I think commit 0198ad917e7d8612cd5ae3356fb9115febfe2eac should complete this PR.  Should be OK to merge if the tests pass.

**Comment by tgflynn** [comment]
It looks like the tests are failing due to download problems.

**Comment by staticfloat** [comment]
Thanks for your contribution @tgflynn!

File: test/spawn.jl [hunk 3]
```diff
@@ -572,8 +576,15 @@ mktempdir() do dir
...
 
+    end
+
+    # Ensure these tests are done only with a PATH of known contents
+    withenv("PATH" => "$(dir)") do
         # Test that completely missing files also return nothing
         @test Sys.which("this_is_not_a_command") === nothing
+
+        # Check that which behaves correctly when passed a blank string
+        @test isnothing(Base.Sys.which(" "))
```
Comment (StefanKarpinski) [inline test/spawn.jl]: I'm a bit late to the party, but should we also test that when there actually is an executable named ` ` in the path we can find it?

Comment (tgflynn) [reply test/spawn.jl]: There are existing tests for that.  I just added a test for a blank string and made sure that the tests that should return nothing are independent of what's on the filesystem.
